executor/src/subscriber.rs

```diff
@@ -1,95 +1,103 @@
+use std::time::Duration;
 use store::Store;
-use tokio::{sync::watch, task::JoinHandle};
-use tracing::debug;
-use types::{metered_channel, BatchDigest, ReconfigureNotification, SerializedBatchMessage};
+use tokio::{
+    sync::{oneshot, watch},
+    task::JoinHandle,
+};
+use tracing::{debug, error};
+use types::{metered_channel, Batch, BatchDigest, ReconfigureNotification};
 
 #[cfg(test)]
 #[path = "tests/subscriber_tests.rs"]
 pub mod subscriber_tests;
 
 /// The `Subscriber` receives certificates sequenced by the consensus and waits until the
-/// `BatchLoader` downloaded all the transactions references by the certificates; it then
+/// downloaded all the transactions references by the certificates; it then
 /// forward the certificates to the Executor Core.
 pub struct Subscriber {
     /// The temporary storage holding all transactions' data (that may be too big to hold in memory).
-    store: Store<BatchDigest, SerializedBatchMessage>,
+    store: Store<BatchDigest, Batch>,
     /// Receive reconfiguration updates.
     rx_reconfigure: watch::Receiver<ReconfigureNotification>,
     /// A channel to receive consensus messages.
     rx_consensus: metered_channel::Receiver<ConsensusOutput>,
-    /// A channel to the batch loader to download transaction's data.
-    tx_batch_loader: metered_channel::Sender<ConsensusOutput>,
     /// A channel to send the complete and ordered list of consensus outputs to the executor. This
     /// channel is used once all transactions data are downloaded.
     tx_executor: metered_channel::Sender<ConsensusOutput>,
+    // A channel to send commands to the block waiter to receive
+    // a certificate's batches (block).
+    tx_get_block_commands: metered_channel::Sender<BlockCommand>,
+    // When asking for a certificate's payload we want to retry until we succeed, unless
+    // some irrecoverable error occurs. For that reason a backoff policy is defined
+    get_block_retry_policy: ExponentialBackoff,
 }
 
 impl Subscriber {
     /// Spawn a new subscriber in a new tokio task.
     #[must_use]
     pub fn spawn(
-        store: Store<BatchDigest, SerializedBatchMessage>,
+        store: Store<BatchDigest, Batch>,
+        tx_get_block_commands: metered_channel::Sender<BlockCommand>,
         rx_reconfigure: watch::Receiver<ReconfigureNotification>,
         rx_consensus: metered_channel::Receiver<ConsensusOutput>,
-        tx_batch_loader: metered_channel::Sender<ConsensusOutput>,
         tx_executor: metered_channel::Sender<ConsensusOutput>,
     ) -> JoinHandle<()> {
+        let get_block_retry_policy = ExponentialBackoff {
+            initial_interval: Duration::from_millis(500),
+            randomization_factor: backoff::default::RANDOMIZATION_FACTOR,
+            multiplier: backoff::default::MULTIPLIER,
+            max_interval: Duration::from_secs(10), // Maximum backoff is 10 seconds
+            max_elapsed_time: None, // Never end retrying unless a non recoverable error occurs.
```

Comment:
If an unrecoverable error happens, we need to panic and stop processing further transactions.